### PR TITLE
DType support for BatchNorm

### PR DESCRIPTION
**Pull request by vchuravy**
cc @Godricly 

depends on https://github.com/dmlc/mshadow/pull/141

### CODE REVIEW

**Comment by Godricly** [comment]
can you make it run using fp16? I tried on m40 but got cudnn fails. while the fp64 case is functional.

File: src/operator/batch_norm-inl.h [hunk 1]
```diff
@@ -131,70 +133,74 @@ class BatchNormOp : public Operator {
     CHECK_EQ(out_data.size(), 3);
     CHECK_EQ(in_grad.size(), 3);
     Stream<xpu> *s = ctx.get_stream<xpu>();
-    Tensor<xpu, 4> data, grad, grad_in;
-    const real_t scale = static_cast<real_t>(out_grad[batchnorm::kOut].shape_[1]) /
-                         static_cast<real_t>(out_grad[batchnorm::kOut].shape_.Size());
+    Tensor<xpu, 4, DType> data, grad, grad_in;
+    const DType scale = DType(
```
Comment (Godricly) [inline src/operator/batch_norm-inl.h]: The computation type problem remains here.

Comment (vchuravy) [inline src/operator/batch_norm-inl.h]: These should be Int, correct?

Comment (Godricly) [inline src/operator/batch_norm-inl.h]: Ye, It should be handled like the one in forward. About the FP16 cudnn BN, those mean and variance are typed in float not as half. I'll download these code and update them with a fix. Maybe you can double check them later. There is some issues with mshadow implementation using FP16 according to my experiment. The fp32 one works fine but fp16 one fails.  can you check that one later after I fixes current issues?

**Comment by Godricly** [comment]
@vchuravy I just did some update on [DTypeBN](https://github.com/Godricly/mxnet/tree/DTypeBN). However there are some issues to be fixed.
* different mean and variance type in fp16 for cudnn and mshadow
* fp16 mshadow BN seems not working

**Comment by vchuravy** [comment]
@Godricly I will look into it again later next week.

**Comment by vchuravy** [comment]
I won't have time to work on this, anytime soon.
